pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java

```diff
@@ -0,0 +1,432 @@
+    _columnMetadataMap = new HashMap<>();
+    PropertiesConfiguration segmentMetadataPropertiesConfiguration = getPropertiesConfiguration(indexDir);
+    _schema = new Schema();
+
+    init(segmentMetadataPropertiesConfiguration);
+    File creationMetaFile = SegmentDirectoryPaths.findCreationMetaFile(indexDir);
+    if (creationMetaFile != null) {
+      loadCreationMeta(creationMetaFile);
+    }
+
+    setTimeInfo(segmentMetadataPropertiesConfiguration);
+    _totalDocs = segmentMetadataPropertiesConfiguration.getInt(Segment.SEGMENT_TOTAL_DOCS);
+  }
+
+  /**
+   * For REALTIME consuming segments.
+   */
+  public SegmentMetadataImpl(String rawTableName, String segmentName, Schema schema, long creationTime) {
+    _indexDir = null;
+    _columnMetadataMap = null;
+    _rawTableName = rawTableName;
+    _segmentName = segmentName;
+    _schema = schema;
+    _creationTime = creationTime;
+  }
+
+  public static PropertiesConfiguration getPropertiesConfiguration(File indexDir) {
+    File metadataFile = SegmentDirectoryPaths.findMetadataFile(indexDir);
+    Preconditions.checkNotNull(metadataFile, "Cannot find segment metadata file under directory: %s", indexDir);
+
+    return CommonsConfigurationUtils.fromFile(metadataFile);
+  }
+
+  /**
+   * Helper method to set time related information:
+   * <ul>
+   *   <li> Time column Name. </li>
+   *   <li> Tine Unit. </li>
+   *   <li> Time Interval. </li>
+   *   <li> Start and End time. </li>
+   * </ul>
+   */
+  private void setTimeInfo(PropertiesConfiguration segmentMetadataPropertiesConfiguration) {
+    _timeColumn = segmentMetadataPropertiesConfiguration.getString(Segment.TIME_COLUMN_NAME);
+    if (segmentMetadataPropertiesConfiguration.containsKey(Segment.SEGMENT_START_TIME)
+        && segmentMetadataPropertiesConfiguration.containsKey(Segment.SEGMENT_END_TIME)
+        && segmentMetadataPropertiesConfiguration.containsKey(Segment.TIME_UNIT)) {
+      try {
+        _timeUnit = TimeUtils.timeUnitFromString(segmentMetadataPropertiesConfiguration.getString(Segment.TIME_UNIT));
+        assert _timeUnit != null;
+        _timeGranularity = new Duration(_timeUnit.toMillis(1));
+        String startTimeString = segmentMetadataPropertiesConfiguration.getString(Segment.SEGMENT_START_TIME);
+        String endTimeString = segmentMetadataPropertiesConfiguration.getString(Segment.SEGMENT_END_TIME);
+        _segmentStartTime = Long.parseLong(startTimeString);
+        _segmentEndTime = Long.parseLong(endTimeString);
+        _timeInterval =
+            new Interval(_timeUnit.toMillis(_segmentStartTime), _timeUnit.toMillis(_segmentEndTime), DateTimeZone.UTC);
+      } catch (Exception e) {
+        LOGGER.warn("Caught exception while setting time interval and granularity", e);
+      }
+    }
+  }
+
+  private void loadCreationMeta(File crcFile)
+      throws IOException {
+    if (crcFile.exists()) {
+      final DataInputStream ds = new DataInputStream(new FileInputStream(crcFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *ds*. The resource is referenced at lines 182, 183, 184. The resource is closed at line 185. There are other execution paths that do not contain closure statements, for example, when DataInputStream.readLong throws an exception. Ensure *ds* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).